core/Pos/Crypto/HD.hs

```diff
@@ -0,0 +1,107 @@
+-- | Hierarchical derivation interface.
+
+module Pos.Crypto.HD
+       ( HDPassphrase
+       , HDAddressPayload (..)
+       , packHDAddressAttr
+       , deriveHDPublicKey
+       , deriveHDSecretKey
+       , deriveHDPassphrase
+       ) where
+
+import           Cardano.Crypto.Wallet        (deriveXPrv, deriveXPrvHardened, deriveXPub,
+                                               unXPub)
+import           Crypto.Cipher.ChaChaPoly1305 as C
+import           Crypto.Error                 (CryptoFailable (..))
+import           Crypto.Hash                  (SHA512 (..))
+import qualified Crypto.KDF.PBKDF2            as PBKDF2
+import           Data.ByteArray               as BA (ByteArrayAccess, convert)
+import           Data.ByteString.Char8        as B
+import           Universum
+
+import           Pos.Binary.Class             (encodeStrict)
+import           Pos.Crypto.Signing           (PublicKey (..), SecretKey (..))
+
+-- | Passphrase is a hash of root public key.
+-- We don't use root public key to store money, we use hash of it instead.
+newtype HDPassphrase = HDPassphrase ByteString
+
+-- | HDAddressPayload consists of
+--
+-- * serialiazed and encrypted with symmetric scheme path from the root
+-- key to given descendant key with passphrase (via ChaChaPoly1305 algorithm)
+--
+-- * cryptographic tag
+--
+-- For more information see 'packHDAddressAttr' and 'encryptChaChaPoly'.
+newtype HDAddressPayload = HDAddressPayload ByteString
+    deriving (Eq, Ord, Show, NFData, Generic)
+
+-- | Compute passphrase as hash of the root public key.
+deriveHDPassphrase :: PublicKey -> HDPassphrase
+deriveHDPassphrase (PublicKey pk) = HDPassphrase $
+    PBKDF2.generate
+        (PBKDF2.prfHMAC SHA512)
+        (PBKDF2.Parameters
+             500 -- Parameters for the hashing function. 500 iter of PBDKF2 with HMAC-SHA256
+             passLen)
+        (unXPub pk)
+        ("address-hashing"::ByteString)
+  where
+    -- Password length in bytes
+    passLen = 32
+
+-- Direct children of node are numbered from 0 to 2^32-1.
+-- Child with index less or equal @maxHardened@ is a hardened child.
+maxHardened :: Word32
+maxHardened = 2 ^ (31 :: Word32) - 1
+
+-- | Derive public key from public key in non-hardened (normal) way.
+-- If you try to pass index more than @maxHardened@, error will be called.
+deriveHDPublicKey :: PublicKey -> Word32 -> PublicKey
+deriveHDPublicKey (PublicKey xpub) childIndex
+    -- Is it the best solution?
+    | childIndex <= maxHardened = error "Wrong index for non-hardened derivation"
+    | otherwise = PublicKey $ deriveXPub xpub (childIndex - maxHardened - 1)
+
+-- | Derive secret key from secret key.
+-- If @childIndex <= maxHardened@ key will be deriving hardened way, otherwise non-hardened.
+deriveHDSecretKey :: ByteArrayAccess passPhrase
+                  => passPhrase
```

Comment:
@pva701 Since you accept `passPhrase`, I think you should work with `EncryptedSecretKey`, not `SecretKey`.
`SecretKey` is a key created / used with empty passphrase